src/rnnlm/arpa-sampling.cc

```diff
@@ -0,0 +1,145 @@
+// arpa-sampling.cc
+
+#include <algorithm>
+#include <iostream>
+#include <iterator>
+#include <math.h>
+#include <string>
+
+#include "arpa-sampling.h"
+
+namespace kaldi {
+
+// This function reads in each ngram line from an ARPA file
+void ArpaSampling::ConsumeNGram(const NGram& ngram) {
+  int32 cur_order = ngram.words.size();
+  int32 word = ngram.words.back(); // word is the last word in a ngram term 
+  HistType history(ngram.words.begin(), ngram.words.begin() + cur_order - 1);
+  KALDI_ASSERT(history.size() == cur_order - 1);
+
+  // log probability of a ngram term (natural log)
+  BaseFloat log_prob = ngram.logprob;
+  // backoff log probability of a ngram term (natural log)
+  BaseFloat backoff_weight = ngram.backoff;
+  std::pair <BaseFloat, BaseFloat> probs_pair;
+  probs_pair = std::make_pair(log_prob, backoff_weight);
+  probs_[cur_order - 1][history].insert({word, probs_pair});
+ 
+  // get the total number of words from an ARPA file
+  if (cur_order == 1) {
+    num_words_++;
+  }
+}
+
+void ArpaSampling::HeaderAvailable() {
+  ngram_counts_ = NgramCounts();
+  ngram_order_ = NgramCounts().size(); 
+  probs_.resize(ngram_order_);
+}
+
+BaseFloat ArpaSampling::GetProb(int32 order, int32 word, const HistType &history) {
+  BaseFloat prob = 0.0;
+  NgramType::const_iterator it = probs_[order - 1].find(history);
+  if (it != probs_[order - 1].end() &&
+      probs_[order-1][history].find(word) != probs_[order-1][history].end()) {
+    prob += probs_[order-1][history][word].first;
+  } else { // backoff to the previous order
+    order--;
+    if (order >= 1) {
+      HistType::const_iterator first = history.begin() + 1;
+      HistType::const_iterator last = history.end();
+      HistType h(first, last);
+      prob += GetProb(order, word, h);
+      int32 word_new = history.back();
+      HistType::const_iterator last_new = history.end() - 1;
+      HistType h_new(history.begin(), last_new);
```

Comment:
Every time you construct a vector, the *object* itself is constructed on the stack, so is free in memory sense. The vector object contains  essentially a pointer to data in memory and vector's size. But this *data* is always allocated with the operator new on the heap. Even those small allocations that freed soon lead to memory fragmentation, and are potentially expensive at runtime. `h_new` is the `h` but without the last element, and `h` is not needed any more. Instead, remove the last element of `h` and get rid of `h_new`.  `h.pop_back()` will do the trick.